jiant/tasks/lib/bucc2018.py

```diff
@@ -0,0 +1,424 @@
+    return scores
+
+
+def score(x, y, fwd_mean, bwd_mean, margin, dist="cosine"):
+    # Adapted From: https://github.com/google-research/xtreme/blob/
+    #               522434d1aece34131d997a97ce7e9242a51a688a/third_party/utils_retrieve.py
+    if dist == "cosine":
+        return margin(x.dot(y), (fwd_mean + bwd_mean) / 2)
+    else:
+        l2 = ((x - y) ** 2).sum()
+        sim = 1 / (1 + l2)
+        return margin(sim, (fwd_mean + bwd_mean) / 2)
+
+
+def knn(x, y, k, use_gpu, dist="cosine"):
+    # Adapted From: https://github.com/google-research/xtreme/blob/
+    #               522434d1aece34131d997a97ce7e9242a51a688a/third_party/utils_retrieve.py
+    return knn_gpu(x, y, k) if use_gpu else knn_cpu(x, y, k, dist)
+
+
+def knn_gpu(x, y, k, mem=5 * 1024 * 1024 * 1024):
+    # Adapted From: https://github.com/google-research/xtreme/blob/
+    #               522434d1aece34131d997a97ce7e9242a51a688a/third_party/utils_retrieve.py
+    import faiss
+
+    dim = x.shape[1]
+    batch_size = mem // (dim * 4)
+    sim = np.zeros((x.shape[0], k), dtype=np.float32)
+    ind = np.zeros((x.shape[0], k), dtype=np.int64)
+    for xfrom in range(0, x.shape[0], batch_size):
+        xto = min(xfrom + batch_size, x.shape[0])
+        bsims, binds = [], []
+        for yfrom in range(0, y.shape[0], batch_size):
+            yto = min(yfrom + batch_size, y.shape[0])
+            idx = faiss.IndexFlatIP(dim)
+            idx = faiss.index_cpu_to_all_gpus(idx)
+            idx.add(y[yfrom:yto])
+            bsim, bind = idx.search(x[xfrom:xto], min(k, yto - yfrom))
+            bsims.append(bsim)
+            binds.append(bind + yfrom)
+            del idx
+        bsims = np.concatenate(bsims, axis=1)
+        binds = np.concatenate(binds, axis=1)
+        aux = np.argsort(-bsims, axis=1)
+        for i in range(xfrom, xto):
+            for j in range(k):
+                sim[i, j] = bsims[i - xfrom, aux[i - xfrom, j]]
+                ind[i, j] = binds[i - xfrom, aux[i - xfrom, j]]
+    return sim, ind
+
+
+def knn_cpu(x, y, k, dist="cosine"):
+    # Adapted From: https://github.com/google-research/xtreme/blob/
+    #               522434d1aece34131d997a97ce7e9242a51a688a/third_party/utils_retrieve.py
+    import faiss
+
+    # x: query, y: database
+    dim = x.shape[1]
+    if dist == "cosine":
+        idx = faiss.IndexFlatIP(dim)
+    else:
+        idx = faiss.IndexFlatL2(dim)
+    idx.add(y)
+    sim, ind = idx.search(x, k)
+
+    if dist != "cosine":
+        sim = 1 / (1 + sim)
+    return sim, ind
+
+
+def get_unique_lines(text_hashes):
+    unique_text_hashes = []
+    unique_indices = []
+    unique_lookup = {}
+    for text_hash in text_hashes:
+        if text_hash not in unique_lookup:
+            unique_lookup[text_hash] = len(unique_lookup)
+            unique_text_hashes.append(text_hash)
+        unique_indices.append(unique_lookup[text_hash])
+    return unique_indices, unique_text_hashes
```

Comment:
Seems needlessly complex (unless I am missing something?). Why not convert `text_hashes` to a set and use the corresponding indices?

Reply:
The important part here is to get `unique_indices`, which runs from 0 to \#unique. If I'm understanding your proposal correctly, I think we would still need to create a dictionary mapping `text_hash->idx`. That said, I don't feel strongly either way and can switch over.

Reply:
The docstring helps. Here is an alternative implementation which may be simpler to read: https://stackoverflow.com/a/34313968

Reply:
My version looks like the first answer on that page :p In that case, let's just move ahead with the current implementation?